### ISSUE TITLE
Junction subdirs in repo are not considered part of the repo

### ISSUE THREAD

**Issue by IkaR49** [git-for-windows/git#2811]
- [ ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.28.0.windows.1
cpu: x86_64
built from commit: 77982caf269b7ee713a76da2bcf260c34d3bf7a7
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 6.3.9600]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Notepad++
Custom Editor Path:
Path Option: Cmd
Plink Path: C:\Program Files\TortoiseGit\bin\TortoiseGitPlink.exe
SSH Option: Plink
Tortoise Option: true
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Probably not.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I test it in Git Bash, cmd and powershell.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
> git clone test-repo
> mkdir test-dir
> cd test-repo
> mklink /j junc-test-dir ..\test-dir
> cd junc-test-dir
> git log
```
 - What did you expect to occur after running these commands?

I expect see log.

 - What actually happened instead?

I see message:
```
fatal: not a git repository (or any of the parent directories): .git
```
But this worked in version 2.19.1 (I can be a little wrong with the version, I don't remember which one was installed earlier).

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

No specific repository.

**Comment by dscho** [comment]
This worked in version v2.19.1 due to a bug that was hence fixed.

That same bug would have allowed `git clean` to delete all the files in the target directory of that junction.

It is related to Git's behavior on Linux where it would not traverse a filesystem boundary when looking for a `.git` directory: If you have mounted a drive under, say, `/mnt/usb-disk/` and called `git rev-parse HEAD` there, and if there was no `/mnt/usb-disk/.git`, it would not even look for `/mnt/.git` because that is a different file system.

Since junctions are more similar to Unix mounts than to Unix symbolic links (after all, it is the NTFS driver that resolves junctions similar to how Linux would resolve mounted directories), when Git looks for the current working directory, it will see the _actual_ working directory, i.e. the target of that junction.

In short: the behavior you are seeing _now_ is much more in line with what Git is designed to do.

**Comment by IkaR49** [comment]
Maybe you should make this behavior configurable? With the default behavior as you described?

**Comment by dscho** [comment]
> Maybe should make this behavior configurable?

If you truly want to have an option to turn this off, I will help you get there, but you will have to do most of the work. You will want to start as described here: https://github.com/git-for-windows/git/wiki/Building-Git

The actual work will have to address two issues, actually:

- roll back the changes of https://github.com/git-for-windows/git/pull/2268 under the config option.
- roll back the changes of https://github.com/git-for-windows/git/pull/2504 under the config option.

As to the first, I think that the [`#ifndef ... #else ... #endif`](https://github.com/git-for-windows/git/pull/2268/files#diff-f17b395741228682b080e923df6e0d4dR176-R193) will need to be turned into a run-time conditional.

As to the second, I think an early return from `mingw_strbuf_realpath()` (guarded under the config option) would need to be added here: https://github.com/git-for-windows/git/pull/2504/files#diff-a0c7aa297d8da1149ce761c31dc59328R1266.

As to the config option, it should probably imitate `core.longPaths` (i.e. you need to look via `git grep -i core.longpaths` for code/documentation/tests that need to be copy-edited).

**Comment by IkaR49** [comment]
I don't have enough time right now because of my work, but I'll get on with it. I think I can do it in an hour in two weeks :)

What do you think: should I offer similar functionality for the linux-git (for mount-points)?

**Comment by dscho** [comment]
> I don't have enough time right now because of my work, but I'll get on with it. I think I can do it in an hour in two weeks :)

If you can do it within an hour... 😁 


> What do you think: should I offer similar functionality for the linux-git (for mount-points)?

Those patches are still only in Git for Windows, not in upstream Git. But yes, I would try to have all platforms that support mount points be treated the same.
